### DISCUSSION TITLE
attiny1616 with both hardware UART and SoftwareSerial in operation?

### DISCUSSION THREAD

**Discussion by celem** [SpenceKonde/megaTinyCore#1084] [Q&A] [answered]
My project uses an attiny1616 with megaTinyCore but I need two UARTS. I would like to use the hardware UART to connect to a remote but nearby device plus SoftwareSerial on a port as TX only to a very close external device. The hardware UART and SoftwareSerial would not be used simultaneously. Baud rates for both serial links are yet to be determined but I have flexibility. Is this possible on the attiny1616 with both hardware UART and SoftwareSerial in operation??

**Comment by hmeijdam** [comment, marked as answer]
Since you do not use the serial links at the same time, did you consider switching between the alternate pin locations of the UART? after RX/TX Closing serial, switching to the alternate pins, beginning serial again, RX/TX to alternate pins, closing serial, switch back, etc...

**Comment by celem** [comment]
Thank you hmeijdam. I had not thought of this as I have never before used _Serial.swap(mux option) or Serial.pins(tx, rx)_. I just gave that section a quick look and I'll have to do some in-depth reading to understand what I will need to do - swap ports, change baud rate, etc.

**Comment by hmeijdam** [comment]
Yes, here is an example. Works simple and you can set the baudrate differently.

```
void setup() {
}

void loop() {
Serial.begin(115200);
Serial.println("This prints on the default UART pins");
Serial.end();
delay(1000);
Serial.swap(1);// change to alternate UART pins
Serial.begin(9600);
Serial.println("This prints on the alternate UART pins");
Serial.end();
delay(1000);
Serial.swap(0);// change to default UART pins
}
```

**Reply by celem** [reply]
hmeijdam - thanks even more. This may save the day on this project!

**Comment by WaynePauly** [comment]
I have the same problem. What would be the minimum latency switching back and forth ( on a AtTiny3216 )?  Does anyone know of a softwareserial library that works on these Tiny's?

**Reply by mechatroniks-git** [reply]
Is there a reason you want to use the 1-Series chips? The 2-Series chips have dual UARTs. ATtiny1624/1626/1627 or ATtiny3224/3226/3227 should do the trick without software serial.

**Reply by WaynePauly** [reply]
Only because I have 3216's but your suggestion might be the best: buy -2 parts and be done with the problems! Thanks.

**Reply by celem** [reply]
While @mechatroniks-git  question wasn't directed to me, I will provide my reason as to why I'm using a ATtiny1616 instead of a 2-series chip. In a nutshell I need the part with the best availability at JLCPCB in China and that is the ATTINY1616-MNR. Almost all  of the 2-series parts  are at or near zero inventory. Yes, I can ship JLCPCB the parts that they lack but cost, time are among my reasons to select only well stocked parts. So, the best part for the job is the obtainable part.
